exoscale/resource_exoscale_nlb.go

```diff
@@ -0,0 +1,275 @@
+			Computed: true,
+		},
+		"state": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"ip_address": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"services": {
+			Type:     schema.TypeSet,
+			Computed: true,
+			Set:      schema.HashString,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+	}
+
+	return &schema.Resource{
+		Schema: s,
+
+		Create: resourceNlbCreate,
+		Read:   resourceNlbRead,
+		Update: resourceNlbUpdate,
+		Delete: resourceNlbDelete,
+		Exists: resourceNlbExists,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(defaultTimeout),
+			Read:   schema.DefaultTimeout(defaultTimeout),
+			Update: schema.DefaultTimeout(defaultTimeout),
+			Delete: schema.DefaultTimeout(defaultTimeout),
+		},
+	}
+}
+
+func resourceNlbCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: beginning create", resourceNlbIDString(d))
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutCreate))
+	defer cancel()
+
+	client := GetComputeClient(meta)
+
+	zone := d.Get("zone").(string)
+
+	nlb, err := client.CreateNetworkLoadBalancer(
+		ctx,
+		zone,
+		&egoscale.NetworkLoadBalancer{
+			Name:        d.Get("name").(string),
+			Description: d.Get("description").(string),
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(nlb.ID)
+
+	log.Printf("[DEBUG] %s: create finished successfully", resourceNlbIDString(d))
+
+	return resourceNlbRead(d, meta)
+}
+
+func resourceNlbRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: beginning read", resourceNlbIDString(d))
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutRead))
+	defer cancel()
+
+	nlb, err := findNlb(ctx, d, meta)
+	if err != nil {
+		return err
+	}
+
+	if nlb == nil {
+		return fmt.Errorf("Network Load Balancer %q not found", d.Id())
+	}
+
+	log.Printf("[DEBUG] %s: read finished successfully", resourceNlbIDString(d))
+
+	client := GetComputeClient(meta)
+
+	return resourceNlbApply(ctx, client, d, nlb)
+}
+
+func resourceNlbExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutRead))
+	defer cancel()
+
+	nlb, err := findNlb(ctx, d, meta)
+	if err != nil {
+		return false, err
+	}
+
+	if nlb == nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func resourceNlbUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: beginning update", resourceNlbIDString(d))
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutUpdate))
+	defer cancel()
+
+	client := GetComputeClient(meta)
+
+	nlb, err := findNlb(ctx, d, meta)
+	if err != nil {
+		return err
+	}
+
+	if nlb == nil {
+		return fmt.Errorf("Network Load Balancer %q not found", d.Id())
+	}
+
+	d.Partial(true)
```

Comment:
No need for partials here, the NLB is updated atomically.

Reply:
246cbf5